File: binary-search-for-the-answer/BinarySearchForAnswer.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class BinarySearchForAnswer {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int N = scanner.nextInt();
        int k = scanner.nextInt();
        List<Integer> numbers = new ArrayList<>(N);
        for (int i = 0; i < N; i ++)
            numbers.add(scanner.nextInt());

        System.out.println(solve(numbers, k));
    }

    static boolean check(int x, int k, List<Integer> array) {
        int line = 1;
        int last_line = array.get(0);
        for (Integer number : array) {
            if (number - last_line >= x) {
                line++;
                last_line = number;
            }
        }
        return line > k;
    }

    static int solve(List<Integer> array, int k) {
        int left = 0;
        int right = array.get(array.size()-1) - array.get(0) + 1;
        while (right - left > 1) {
            int middle = (left + right) / 2;
            if (check(middle, k, array)) {
                left = middle;
            } else {
                right = middle;
            }
        }
        return left;
    }


}

File: binary-search-for-the-answer/binary-search-for-answer.py

def binary_search_for_answer(array, segment_count):
    left = 0
    right = coords[N-1] - coords[0] + 1
    while (right - left > 1):
        mid = (left + right) // 2
        if (check(mid, k, coords)):
         left = mid
        else:
         right = mid
    return left
    # for _ in range(array_count):
        # print(binary_search(array, int(input())))


def main():
    array_count = int(input())
    segment_count = int(input())
    array = [int(input()) for _ in range(array_count)]
    binary_search_for_answer(array, segment_count)

if __name__ == '__main__':
    binary_search_for_answer()